pkg/dynamicaccess/controller.go

```diff
@@ -1,54 +1,210 @@
 package dynamicaccess
 
 import (
+	"context"
 	"crypto/ecdsa"
+	"time"
 
+	"github.com/ethersphere/bee/v2/pkg/file/loadsave"
+	"github.com/ethersphere/bee/v2/pkg/file/pipeline"
+	"github.com/ethersphere/bee/v2/pkg/file/pipeline/builder"
+	"github.com/ethersphere/bee/v2/pkg/file/redundancy"
+	"github.com/ethersphere/bee/v2/pkg/kvs"
 	kvsmock "github.com/ethersphere/bee/v2/pkg/kvs/mock"
+	"github.com/ethersphere/bee/v2/pkg/storage"
 	"github.com/ethersphere/bee/v2/pkg/swarm"
 )
 
+type GranteeManager interface {
+	//PUT /grantees/{grantee}
+	//body: {publisher?, grantee root hash ,grantee}
+	Grant(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error
+	//DELETE /grantees/{grantee}
+	//body: {publisher?, grantee root hash , grantee}
+	Revoke(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error
+	//[ ]
+	//POST /grantees
+	//body: {publisher, historyRootHash}
+	Commit(ctx context.Context, granteesAddress swarm.Address, actRootHash swarm.Address, publisher *ecdsa.PublicKey) (swarm.Address, swarm.Address, error)
+
+	//Post /grantees
+	//{publisher, addList, removeList}
+	HandleGrantees(ctx context.Context, rootHash swarm.Address, publisher *ecdsa.PublicKey, addList, removeList []*ecdsa.PublicKey) error
+
+	//GET /grantees/{history root hash}
+	GetGrantees(ctx context.Context, rootHash swarm.Address) ([]*ecdsa.PublicKey, error)
+}
+
+// TODO: ądd granteeList ref to history metadata to solve inconsistency
 type Controller interface {
-	DownloadHandler(timestamp int64, enryptedRef swarm.Address, publisher *ecdsa.PublicKey, tag string) (swarm.Address, error)
-	UploadHandler(ref swarm.Address, publisher *ecdsa.PublicKey) (swarm.Address, error)
+	GranteeManager
+	DownloadHandler(ctx context.Context, timestamp int64, encryptedRef swarm.Address, publisher *ecdsa.PublicKey, historyRootHash swarm.Address) (swarm.Address, error)
+	UploadHandler(ctx context.Context, reference swarm.Address, publisher *ecdsa.PublicKey, historyRootHash *swarm.Address) (swarm.Address, swarm.Address, swarm.Address, error)
 }
 
-type defaultController struct {
-	history        History
-	granteeManager GranteeManager
-	accessLogic    ActLogic
+type controller struct {
+	accessLogic ActLogic
+	granteeList GranteeList
+	//[ ]: do we need to protect this with a mutex?
```

Comment:
Should this be a TODO or it can be tracked with the "[ ]:" prefix?

Reply:
Gratnee related functions will be handled in a later PR.